### PR TITLE
Fulltext api deprecation

### PR DESCRIPTION
**Pull request by angrykoala**
# Description

Top level fulltext queries are now deprecated due to a poorly defined design on our side regarding future API and fulltext.

The `fulltext` field has been deprecated for "simple" queries. Being now supported on connections and aggregation queries.

Arguments have also being deprecated, in favour of an `index` and `name` arguments:

```graphql
query {
  moviesConnection(fulltext: { phrase: "My search", index: MovieTitle }) {
    edges {
      node {
        title
      }
    }
  }
}
```

Complexity: Medium

This PR does not add `score` into the connection query, effectively making the new API a regression until this is implemented in a different PR

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Looks like you've updated the documentation!

Check out your changes at https://neo4j-graphql-docs-3110.surge.sh

**Comment by neo4j-team-graphql** [comment]
Made some changes to the GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at https://neo4j-graphql-toolbox-stage-3110.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3403 | 3403 | 46 | 46 | 1134 |
|  aggregations.NestedAggregation | 14551 | 14551 | 83 | 83 | 2174 |
|  aggregations.AggregationWithWhere | 11979 | 11979 | 53 | 57 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116987 | 22116987 | 3253 | 3310 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116987 | 22116987 | 2909 | 2915 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9734 | 9734 | 47 | 51 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9937 | 9937 | 41 | 53 | 2174 |
|  aggregations.DeeplyNestedCount | 13070331 | 13070329 | 4419 | 4626 | 2008534 |
|  batch-create.BatchCreate | 4200 | 4200 | 121 | 170 | 600 |
|  connect.createAndConnect | 14424 | 14424 | 162 | 222 | 3003 |
|  connections.Connection | 14082 | 14082 | 64 | 59 | 2174 |
|  connections.NestedConnection | 41462 | 41462 | 110 | 123 | 4516 |
|  connections.ConnectionWithSort | 2282 | 2282 | 69 | 93 | 1040 |
|  connections.ConnectionWithSortAndCypher | 2282 | 2282 | 165 | 290 | 1040 |
|  create.SimpleMutation | 7 | 7 | 53 | 68 | 1 |
|  cypher-directive.TopLevelCypherDirective | 116092 | 116092 | 677 | 698 | 12241 |
|  cypher-directive.TopLevelCypherDirectiveWithColumnName | 147609 | 147611 | 164 | 192 | 12241 |
|  delete.SimpleDelete | 19401 | 19401 | 291 | 396 | 1040 |
|  delete.NestedDeleteInUpdate | 22200 | 22200 | 132 | 164 | 2040 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 33 | 40 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22900 | 22900 | 69 | 98 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 3450 | 3450 | 23 | 30 | 1134 |
|  2925.NestedSingleRelationshipRequiredFilter | 9383 | 9383 | 47 | 66 | 1040 |
|  query.SimpleQuery | 15160 | 15160 | 58 | 67 | 2174 |
|  query.QueryWhere | 9678 | 9707 | 45 | 54 | 2156 |
|  query.SimpleQueryWithNestedWhere | 9856 | 9885 | 54 | 57 | 2156 |
|  query.Nested | 10092037 | 10092037 | 9690 | 10537 | 2008534 |
|  query.NestedWithFilter | 40401 | 40401 | 94 | 96 | 4511 |
|  query.OrFilterOnRelationships | 41132 | 43809 | 157 | 172 | 1252 |
|  query.OrFilterOnRelationshipsAndNested | 35230 | 37303 | 242 | 231 | 1252 |
|  query.QueryWithNestedIn | 13358 | 14117 | 76 | 63 | 1202 |
|  query.NestedConnectionWhere | 9834 | 9834 | 82 | 80 | 2174 |
|  query.DeeplyNestedConnectionWhere | 9931 | 10089 | 74 | 84 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 19212 | 19308 | 134 | 146 | 1620 |
|  query.NestedWithRelationshipSingleFilters | 3024400 | 3024400 | 610 | 566 | 1003942 |
|  query.Fulltext | 96 | 96 | 25 | 28 | 16 |
|  query.FulltextWithNestedQuery | 587 | 587 | 41 | 115 | 84 |
|  sorting.SortMultipleTypes | 2513 | 2513 | 62 | 144 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 1416 | 1488 | 168 | 297 | 1040 |
|  sorting.SortOnNestedFields | 14082 | 14082 | 51 | 70 | 2174 |
|  sorting.SortDeeplyNestedFields | 42196 | 42196 | 83 | 113 | 4516 |
|  sorting.SortWithTopLevelCypher | 2119 | 2119 | 98 | 163 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 66 | 74 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 70 | 71 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 309 | 315 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 278 | 287 | 33033 |
|  update.NestedUpdate | 2119 | 2119 | 89 | 69 | 1040 |

</details>
        
Old Schema Generation: 31.990s
Schema Generation: 30.265s
Old Subgraph Schema Generation: 43.802s
Subgraph Schema Generation: 47.281s

File: packages/graphql/src/schema/augment/fulltext.ts [hunk 1]
```diff
@@ -17,50 +17,114 @@
...
 
+function createEnumType(name: string, rawValues: string[]): GraphQLEnumType {
+    const values = rawValues.reduce((acc, value) => {
+        return {
+            ...acc,
+            [value]: {
+                value,
+            },
+        };
+    }, {} as GraphQLEnumValueConfigMap);
```
Comment (mjfwebb) [inline packages/graphql/src/schema/augment/fulltext.ts:51]:
```suggestion
    const values = rawValues.reduce((acc: GraphQLEnumValueConfigMap, value) => {
        return {
            ...acc,
            [value]: {
                value,
            },
        };
    }, {});
```

Just a little nicer not to use `as` in my opinion :)

File: packages/graphql/src/schema/augment/fulltext.ts [hunk 2]
```diff
@@ -17,50 +17,114 @@
...
     composer: SchemaComposer,
     nodeWhereTypeName: string,
     nodeSortTypeName: string
 ) {
     if (node.fulltextDirective) {
-        const fields = node.fulltextDirective.indexes.reduce((res, index) => {
+        const indexes = node.fulltextDirective.indexes.map((index) => {
+            const indexName = index.indexName || index.name;
+            if (indexName === undefined) {
+                throw new Error("The name of the fulltext index should be defined using the indexName argument.");
```
Comment (mjfwebb) [inline packages/graphql/src/schema/augment/fulltext.ts:69]: Would it make sense to throw a `Neo4jGraphQLError` here?

File: packages/graphql/src/schema/augment/fulltext.ts [hunk 3]
```diff
@@ -17,50 +17,114 @@
...
+                },
             };
         }, {});
 
         const fulltextResultDescription = `The result of a fulltext search on an index of ${node.name}`;
         const fulltextWhereDescription = `The input for filtering a fulltext query on an index of ${node.name}`;
         const fulltextSortDescription = `The input for sorting a fulltext query on an index of ${node.name}`;
 
+        // Sets a default index if there is only one index available
+        let defaultIndex = undefined as any;
```
Comment (mjfwebb) [inline packages/graphql/src/schema/augment/fulltext.ts:111]:
```suggestion
        let defaultIndex: string | undefined = undefined;
```

Let's make sure we don't add `any` if possible

**Review by mjfwebb — COMMENTED** [review]
Just some comments. I'll give this another look tomorrow :)

File: packages/graphql/src/schema/augment/fulltext.ts (hunk 1 above)
Comment (angrykoala) [reply packages/graphql/src/schema/augment/fulltext.ts:51]: I tend to prefer strictly typing `reduce` functions, not having the `as` means that that reduce function now infers value to be an object (and an empty one at that!) as `reduce` uses the type of the second parameter to infer types

This means, that if I mess my code like:

```
    const values = rawValues.reduce((acc, value) => {
        return {
            ...acc,
            tst: "dsa",
            [value]: {
                value,
            },
        };
    }, {});
```

The typing system will not detect the problem, and furthermore, the returned `values` is still of type empty object so it gets silently passed without a problem.

File: packages/graphql/src/schema/augment/fulltext.ts (hunk 2 above)
Comment (angrykoala) [reply packages/graphql/src/schema/augment/fulltext.ts:69]: Maybe? We don't have a standard for errors sadly. This one is more of a fatal error I think, as the graphql typings should avoid this case, so I'm not sure if `Neo4jGraphQLError` is the appropriate type, but happy to change it if you think is better

**Pull request closed without merging**